lib/install/rust.js

```diff
@@ -0,0 +1,559 @@
+      }
+      config.stagingDir = check.path;
+
+      return exports.checkRustlib(config.rustv);
+    })
+    .then(check => {
+      if (!check.exists) {
+        throw new Error(`MIPS libstd v${config.rustv} not installed.`);
+      }
+      config.rustlibPath = check.path;
+
+      return exports.toolchainPath();
+    })
+    .then(toolchainPath => {
+      config.toolchainPath = toolchainPath;
+
+      return config;
+    });
+};
+
+// Confirms that the user has a version of rustc and cargo installed. Resolves
+// with the current rustc version, rejects if either executable is not found
+// on the system.
+module.exports.rustVersion = () => {
+  // Check that rustc exists and get its version.
+  return new Promise((resolve, reject) => {
+    var rustc = cp.spawn('rustc', ['-V']);
+    var rustcOut = [];
+    rustc.stdout.on('data', (data) => {
+      rustcOut.push(data);
+    });
+    rustc.on('error', reject);
+    rustc.on('close', (status) => {
+      var out = Buffer.concat(rustcOut).toString();
+      var matches = out.match(/^rustc\s+(\S+)/);
+      if (status !== 0 || !matches) {
+        return reject(new Error('Could not find a locally installed version of `rustc`.'));
+      }
+
+      var version = matches[1];
+
+      // Check if cargo exists also.
+      var cargo = cp.spawn('cargo', ['-V']);
+      cargo.on('error', reject);
+      cargo.on('close', (status) => {
+        if (status !== 0) {
+          return reject(new Error('Could not find a locally installed version of `cargo`'));
+        }
+
+        resolve(version);
+      });
+    });
+  });
+};
+
+// Check the targets of a cargo crate.
+module.exports.cargoMetadata = (destdir) => {
+  return new Promise((resolve, reject) => {
+    var cargo = cp.spawn('cargo', ['metadata', '--no-deps'], {
+      stdio: ['ignore', 'pipe', 'inherit'],
+      cwd: destdir || process.cwd(),
+    });
+    var stdout = [];
+    cargo.stdout.on('data', (data) => {
+      stdout.push(data);
+    });
+    cargo.on('close', (status) => {
+      if (status !== 0) {
+        reject(new Error('Could not inspect project metadata.'));
+      } else {
+        resolve(JSON.parse(Buffer.concat(stdout).toString()));
+      }
+    });
+  });
+};
+
+module.exports.buildTessel = (config) => {
+  var env = Object.assign({}, process.env);
+  Object.assign(env, {
+    STAGING_DIR: config.stagingDir,
+    RUST_TARGET_PATH: config.rustlibPath,
+    PATH: `${path.join(config.toolchainPath, 'bin')}:${env.PATH}`,
+    RUSTFLAGS: `-L ${config.rustlibPath}`,
+  });
+
+  return new Promise((resolve) => {
+    var cargo = cp.spawn('cargo', ['build', '--target=tessel2', '--bin', config.name, '--release'], {
+      env: env,
+      pwd: config.path || process.cwd(),
+      stdio: ['ignore', 'inherit', 'inherit'],
+    });
+
+    cargo.on('error', (error) => {
+      log.error(`${error.stack}`);
+    });
+
+    cargo.on('close', (code) => {
+      if (code !== 0) {
+        process.exit(code);
```

Comment:
@HipsterBrown this and any other non-standard exit will have to be refactored to use the same "safely close all tessels" logic used elsewhere.